File: stock.py
# -*- coding: utf-8 -*-

#title:
#1、一个数组表示股票每天的价格, 只可以进行一次买卖操作并且只能先买后卖, 求何时买卖利润最大
#2、可以买卖无穷次, 买之前必须把以前的股票卖掉, 求最大获利
#3、可以买卖两次, 买之前必须把以前的股票卖掉, 求最大获利

#1、时间复杂度 O(n)
#2、时间复杂度 O(n)
#3、时间复杂度 O(n)


def max_profit1(prices):
    if len(prices) ==0 or len(prices) == 1:
        return 0
    buy = prices[0]
    profit = 0
    for v in prices:
        buy = min(buy, v)
        profit = max(v-buy, profit)
    return profit


def max_profit2(prices):
    if len(prices) ==0 or len(prices) == 1:
        return 0
    profit = 0
    for i in range(len(prices)-1):
        if prices[i]<prices[i+1]:
            profit += prices[i+1]-prices[i]
    return profit


def max_profit3(prices):
    if len(prices) ==0 or len(prices) == 1:
        return 0
    sum = 0
    buy = prices[0]
    sell = prices[-1]
    left = []
    right = []
    profit = 0
    for v in prices:
        buy = min(buy, v)
        profit = max(v - buy, profit)
        left.append(profit)
    profit = 0
    for v in prices[::-1]:
        sell = max(sell, v)
        profit = max(sell - v, profit)
        right.append(profit)
    for i in range(len(prices)):
        if left[i]+right[len(prices)-i-1] > sum:
            sum = left[i]+right[len(prices)-i-1]
    return sum

if __name__ == "__main__":
    prices = [2,1,2,4,3,10]
    print max_profit1(prices)
    print max_profit2(prices)
    print max_profit3(prices)

File: stock.cpp
/*
 *  Best Time to Buy and Sell Stock Problem Collection 
 */

#include <vector>
#include <iostream>
#include <cmath>
#include <climits>

using namespace std;

// Best Time to Buy and Sell Stock 
int MaxProfit(const vector<int> &prices) { 
    int min_price = INT_MAX;
    int max_profit = 0;

    for (size_t i = 0; i < prices.size(); ++i) {
        min_price = min(prices[i], min_price);
        max_profit = max(max_profit, prices[i] - min_price);
    }
    return max_profit;
}

// Best Time to Buy and Sell Stock II 
int MaxProfit2(const vector<int> &prices) {
    int max_profit = 0;

    for (size_t i = 1; i < prices.size(); ++i) {
        max_profit += max(0, prices[i] - prices[i - 1]); 
    }
    return max_profit;
}

// Best Time to Buy and Sell Stock III
int MaxProfit3(const vector<int> &prices) {
    if (prices.size() < 2)
        return 0;

    int global[2][2] = {0, 0, 0, 0};
    int local[2] = {0, 0};

    int prev = 0;
    int cur = 1;
    int min_price = prices[0];

    for (size_t i = 1; i < prices.size(); ++i) {
        int diff = prices[i] - prices[i - 1];
        min_price = min(prices[i], min_price);
        local[cur] = max(global[prev][0] + max(0, diff), local[prev] + diff); 
        global[cur][0] = max(global[cur][0], prices[i] - min_price);
        global[cur][1] = max(global[prev][1], local[cur]);
        swap(prev, cur);
    }
    return global[prev][1];
}

// Best Time to Buy and Sell Stock IV
int MaxProfit4(int k, const vector<int> &prices) {
    if (prices.size() < 2 || k < 1)
        return 0;

    if (k > prices.size() / 2) {
        int ret = 0;
        for (size_t i = 1; i < prices.size(); ++i) {
            ret += max(prices[i] - prices[i - 1], 0);
        }
        return ret;
    }

    vector<int> global[2] = {vector<int>(k, 0), vector<int>(k, 0)};
    vector<int> local[2] = {vector<int>(k, 0), vector<int>(k, 0)};

    int prev = 0;
    int cur = 1;

    for (size_t i = 1; i < prices.size(); ++i) {
        int diff = prices[i] - prices[i - 1];

        for (int j = 0; j < k; ++j) {
            if (j == 0)
                local[cur][j] = max(max(diff, 0), local[prev][j] + diff);
            else
                local[cur][j] = max(global[prev][j - 1] + max(diff, 0),
                                    local[prev][j] + diff);
            global[cur][j] = max(global[prev][j], local[cur][j]);
        }
        swap(prev, cur);
    }

    return global[prev][k - 1];
}

int main(void) {
    vector<int> prices{2, 4, 6, 9, 1, 3, 11, 7};
    cout << MaxProfit(prices) << endl;
    cout << MaxProfit2(prices) << endl;
    cout << MaxProfit3(prices) << endl;
    cout << MaxProfit4(2, prices) << endl;
    return 0;
}
